driver-core/src/main/com/mongodb/internal/operation/MixedBulkWriteOperation.java

```diff
@@ -292,9 +271,10 @@ private BulkWriteResult executeBulkWriteBatch(final RetryState retryState, final
                 currentBulkWriteTracker = BulkWriteTracker.attachNext(retryState, currentBatch);
                 currentBatch = currentBulkWriteTracker.batch().orElseThrow(Assertions::fail);
             } catch (MongoException exception) {
-                if (!(retryState.isFirstAttempt() || (exception instanceof MongoWriteConcernWithResponseException))) {
+                if (!retryState.isFirstAttempt() && !(exception instanceof MongoWriteConcernWithResponseException)) {
```

Comment:
Changed this because I found the new form easier to understand (I was trying to understand what's going on, not very successfully, though).

Reply:
Should there be a separate catch for `MongoWriteConcernWithResponseException` to make this easier to understand?

Reply:
Currently we have
```java
} catch (MongoException exception) {
    if (!retryState.isFirstAttempt() && !(exception instanceof MongoWriteConcernWithResponseException)) {
        addRetryableWriteErrorLabel(exception, maxWireVersion);
    }
    handleMongoWriteConcernWithResponseException(retryState, false);
    throw exception;
}
```
with the change you propose we will have
```java
} catch (MongoWriteConcernWithResponseException exception) {
    handleMongoWriteConcernWithResponseException(retryState, false);
    throw exception;
} catch (MongoException exception) {
    if (!retryState.isFirstAttempt()) {
        addRetryableWriteErrorLabel(exception, maxWireVersion);
    }
    handleMongoWriteConcernWithResponseException(retryState, false);
    throw exception;
}
```

The current code is shorter and does not have code duplication, the proposed code does not have the `instanceof` check. To me the current approach seems the lesser of two evils.